extensions/Gemini/python/src/aiconfig_extension_gemini/Gemini.py

```diff
@@ -0,0 +1,295 @@
+        if options is not None and options.stream is not None:
+            stream = options.stream
+        elif "stream" in completion_data:
+            stream = completion_data["stream"]
+        completion_data["stream"] = stream
+
+        response = await self.model.generate_content_async(**completion_data)
+
+        outputs = None
+        if stream:
+            outputs = await construct_stream_outputs(response, options)
+        else:
+            outputs = construct_regular_outputs(response)
+
+        prompt.outputs = outputs
+        await aiconfig.callback_manager.run_callbacks(
+            CallbackEvent("on_run_complete", __name__, {"result": prompt.outputs})
+        )
+        return prompt.outputs
+
+    def get_output_text(
+        self,
+        prompt: Prompt,
+        aiconfig: "AIConfigRuntime",
+        output: Optional[Output] = None,
+    ) -> str:
+        if not output:
+            output = aiconfig.get_latest_output(prompt)
+
+        if not output:
+            return ""
+
+        if output.output_type == "execute_result":
+            output_data = output.data
+            if isinstance(output_data, str):
+                return output_data
+            else:
+                raise Exception("Not Implemented")
+
+        else:
+            return ""
+
+    def _construct_chat_history(
+        self, prompt: Prompt, aiconfig: "AIConfigRuntime", params: Dict
+    ) -> List:
+        """
+        Constructs the chat history for the model
+        """
+        messages = []
+        # Default to always use chat context
+        remember_chat_context = not hasattr(
+            prompt.metadata, "remember_chat_context"
+        ) or (
+            hasattr(prompt.metadata, "remember_chat_context")
+            and prompt.metadata.remember_chat_context != False
+        )
+        if remember_chat_context:
+            # handle chat history. check previous prompts for the same model. if same model, add prompt and its output to completion data if it has a completed output
+            for i, previous_prompt in enumerate(aiconfig.prompts):
+                # include prompts upto the current one
+                if previous_prompt.name == prompt.name:
+                    break
+
+                previous_prompt_is_same_model = aiconfig.get_model_name(
+                    previous_prompt
+                ) == aiconfig.get_model_name(prompt)
+                if previous_prompt_is_same_model:
+                    previous_prompt_template = resolve_prompt(
+                        previous_prompt, params, aiconfig
+                    )
+                    previous_prompt_output = aiconfig.get_latest_output(previous_prompt)
+                    previous_prompt_output_text = self.get_output_text(
+                        previous_prompt, aiconfig, previous_prompt_output
+                    )
+
+                    messages.append(
+                        {"role": "user", "parts": [{"text": previous_prompt_template}]}
```

Comment:
Just double-checking lowercase role works instead of USER and MODEL ?

Reply:
Not sure if it matters, the palm equivalent was `author: 0`, `author: 1` but the api/model only checked that it was alternating. Fairly certain that the Gemini is doing something similar.

the lowercase user and model worked for this